### PR TITLE
New fast-jar packaging format

### PR DESCRIPTION
**Pull request by stuartwdouglas**
This new format is now created in a subdirectory in the target dir.

This also introduces a new faster ClassLoader than improves startup
time by around 10%.

This is currently in draft state, the implementation has been done however there is a lot more work that needs to happen around this such as documentation and docker file changes. 

We also need to decide if we want to make this the default in the next release, or consider doing a staged release where we include this but don't make it the default so people can try it out, then make it the default at a later date.

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
@gsmet I have changed this so it is not the default, so it should be ready to go in now (assuming CI passes), as it will only affect apps that have explicitly been configured to use this new format.

**Comment by geoand** [comment]
This doesn't work with `container-image-jib`. Should I work on making it work and push into this PR?

File: independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        String packageName = getPackageNameFromClassName(name);
+        ClassLoadingResource[] resources;
+        if (packageName == null) {
+            resources = resourceDirectoryMap.get("");
+        } else {
+            String dirName = packageName.replace(".", "/");
+            resources = resourceDirectoryMap.get(dirName);
+        }
+        if (resources == null) {
+            throw new ClassNotFoundException();
```
Comment (geoand) [inline independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java:48]: Shouldn't this be `throw new ClassNotFoundException(name);`?

File: independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            definePackage(packageName, resources);
+            return defineClass(name, data, 0, data.length,
+                    protectionDomains.computeIfAbsent(resource, new Function<ClassLoadingResource, ProtectionDomain>() {
+                        @Override
+                        public ProtectionDomain apply(ClassLoadingResource ce) {
+                            return ce.getProtectionDomain(RunnerClassLoader.this);
+                        }
+                    }));
+        }
+        throw new ClassNotFoundException();
```
Comment (geoand) [inline independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/RunnerClassLoader.java:65]: Same comment as above

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 3]
```diff
@@ -338,6 +346,148 @@ private JarBuildItem buildThinJar(CurateOutcomeBuildItem curateOutcomeBuildItem,
...
+        Path buildDir = outputTargetBuildItem.getOutputDirectory()
+                .resolve(outputTargetBuildItem.getBaseName());
+        IoUtils.recursiveDelete(buildDir);
+        Files.createDirectories(buildDir);
+
+        //unmodified 3rd party dependencies
+        Path libDir = buildDir.resolve("lib");
+        Files.createDirectories(libDir);
+        //parent first entries
+        Path baseLib = buildDir.resolve("base-lib");
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:368]: Not really important, but maybe `base-lib` should be called `boot-lib` to better reflect what its about? `base-lib` sounds a little to generic

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 4]
```diff
@@ -338,6 +346,148 @@ private JarBuildItem buildThinJar(CurateOutcomeBuildItem curateOutcomeBuildItem,
...
+            generateManifest(runnerZipFs, classPath.toString(), packageConfig, appArtifact, QuarkusEntryPoint.class.getName(),
+                    applicationInfo);
+            Path appInfo = runnerZipFs.getPath(QuarkusEntryPoint.QUARKUS_APPLICATION_DAT);
+            try (OutputStream out = wrapForJDK8232879(Files.newOutputStream(appInfo, DEFAULT_OPEN_OPTIONS))) {
+                ByteArrayOutputStream bs = new ByteArrayOutputStream();
+                SerializedApplication.write(bs, mainClassBuildItem.getClassName(), buildDir, jars);
+                out.write(bs.toByteArray());
+                out.flush();
+            }
+            runnerZipFs.close();
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:484]: This is redundant

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 5]
```diff
@@ -338,6 +346,148 @@ private JarBuildItem buildThinJar(CurateOutcomeBuildItem curateOutcomeBuildItem,
...
+            }
+
+            for (Path resolvedDep : depArtifact.getPaths()) {
+                if (!Files.isDirectory(resolvedDep)) {
+                    if (curateOutcomeBuildItem.getEffectiveModel().getParentFirstArtifacts()
+                            .contains(depArtifact.getKey())) {
+                        final String fileName = depArtifact.getGroupId() + "." + resolvedDep.getFileName();
+                        final Path targetPath = baseLib.resolve(fileName);
+                        Files.copy(resolvedDep, targetPath, StandardCopyOption.REPLACE_EXISTING);
+                        classPath.append(" base-lib/" + fileName);
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:438]:
```suggestion
                        classPath.append(" base-lib/").append(fileName);
```

**Comment by stuartwdouglas** [comment]
I will look at the jib stuff tomorrow, unless you really want to fix it, in which case I would not object :-)

**Comment by geoand** [comment]
I'll take care of it today or tomorrow and push the fix as a commit into the PR if that's OK with you

**Comment by geoand** [comment]
It would be great though if the PR was rebased onto master first and the conflict resolved

**Comment by stuartwdouglas** [comment]
Done

**Comment by geoand** [comment]
Great, thanks

**Comment by geoand** [comment]
I pushed a fix to make Jib work with `fast-jar`

**Comment by geoand** [comment]
Seems like it needs another rebase

**Comment by geoand** [comment]
I'm in favor of getting this is - we really need people to start testing it out. 
For the Dockerfiles, I guess we could have additionally generate  `Dockerfile.jvm.fastjar` or something like that. I don't see us being able to have one dockerfile for both modes

**Comment by stuartwdouglas** [comment]
Once this is in I will look at the Dockerfiles, but it will definitely need a different one.

**Comment by geoand** [comment]
I think my Jib commit got lost. Do you want me to push it into the PR again?

**Comment by geoand** [comment]
I just pushed the jib commit back in

**Comment by geoand** [comment]
My vote is that the new mode shouldn't be the default for the first release, but instead be a staged release.

**Review by geoand — APPROVED** [review]

**Comment by gsmet** [comment]
OK, let's get this one in.

**Comment by gsmet** [comment]
@stuartwdouglas could you write something I could integrate in the announcement so that we get people to test it?

**Comment by wjglerum** [comment]
Any news on an example Dockerfile for the fast-jar packaging? I just broke our deployment when enabling this. Locally the build still worked because I had all the old libraries in the target folder. At the same time I'm also exploring building images with Jib, see https://quarkus.io/guides/container-image

**Comment by geoand** [comment]
The `container-image-jib` extension **does** work today with the new packaging format. For docker, we should at the very least add some comments in the generated Dockerfile telling people what to do when `fast-jar` is being used.

**Comment by wjglerum** [comment]
Thanks, just build an image successfuly indeed. Would be nice to have some instructions indeed.

**Pull request merged**